controller/env/sync.go

```diff
@@ -0,0 +1,73 @@
+/*
+	Copyright NetFoundry, Inc.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package env
+
+import (
+	"github.com/openziti/edge/controller/model"
+	"github.com/openziti/edge/controller/persistence"
+	"github.com/openziti/fabric/controller/network"
+)
+
+// Aliased type for router strategies
+type RouterSyncStrategyType string
+
+// Aliased type for router sync status
+type RouterSyncStatus string
+
+const (
+	RouterSyncNew        RouterSyncStatus = "SYNC_NEW"         //connection accepted but no strategy actions have been taken
+	RouterSyncQueued     RouterSyncStatus = "SYNC_QUEUED"      //connection handed to strategy, but not processed
+	RouterSyncHello      RouterSyncStatus = "SYNC_HELLO"       //connection is beginning hello cycle
+	RouterSyncHelloWait  RouterSyncStatus = "SYNC_HELLO_WAIT"  //hello received from router, but there are too many synchronizing routers
+	RouterSyncResyncWait RouterSyncStatus = "SYNC_RESYNC_WAIT" //router requested a resync, in queue
+	RouterSynInProgress  RouterSyncStatus = "SYNC_IN_PROGRESS" //hello finished, starting to send state
+	RouterSyncDone       RouterSyncStatus = "SYNC_DONE"        //initial state sent
+
+	//Error states
+	RouterSyncUnknown      RouterSyncStatus = "SYNC_UNKNOWN"       //the router is currently unknown
+	RouterSyncDisconnected RouterSyncStatus = "SYNC_DISCONNECTED"  //strategy was disconnected before finishing
+	RouterSyncHelloTimeout RouterSyncStatus = "SYNC_HELLO_TIMEOUT" //sync failed due to a hello timeout
+)
+
+// RouterSyncStrategy handles the life cycle of an Edge Router connecting to the controller, synchronizing
+// any upfront state and then maintaining state after that.
+type RouterSyncStrategy interface {
```

Comment:
Interfaces that can be used to implement composable and reusable pieces. Probably needs some more refinement but that will happen when a new strategy is added.